### PR TITLE
Add detection callback to MaskSecrets, make Detection immutable

### LINKED ISSUE #169: Review code for mutable reference types that may get lost in dictionaries
E.g., our `Detection` class. 

https://learn.microsoft.com/en-us/dotnet/api/system.object.gethashcode?view=net-9.0#notes-to-inheritors

### LINKED ISSUE #172: API to do secret masking and receive detection info in one pass
Currently, you can either use SecretMasker.DetectSecrets or SecretMasker.MaskSecrets. If you use the latter, your input is redacted without returning to you the data about the detections that triggered redactions. 

Consider the scenario where you'd like to perform redaction and log information about the redactions performed. This will be needed in follow-up work after https://github.com/microsoft/azure-pipelines-agent/pull/5154

Possible approaches discussed so far:

1. API that returns both redacted string and detections.
2. API that can use detections to perform redactions.
3. Callback from redaction API with each detection being used for redaction.

### PR DESCRIPTION
**Pull request by nguerrera**
* Add callback to MaskSecrets so detection data can be processed. Fix #172
* Make detection sealed, immutable, and compare equality by reference. Fix #169 
* Optimize away some intermediate allocations in MaskSecrets.

Took on #169 because a mutation in the callback would break things and defending against it would make confusing defensive copies even more confusing.

Then took on the optimization because the algorithm was relying on mutating Detection.Length.

### CODE REVIEW

File: src/Microsoft.Security.Utilities.Core/SecretMasker.cs [hunk 1]
```diff
@@ -122,7 +122,7 @@ public virtual void AddRegex(RegexPattern regexSecret)
         }
     }
 
-    public string MaskSecrets(string input)
+    public string MaskSecrets(string input, Action<Detection>? onDetection = null)
```
Comment (nguerrera) [inline src/Microsoft.Security.Utilities.Core/SecretMasker.cs]: Copilot chose this name, but maybe this isn't so idiomatic for C#. Suggestions? `detectionAction`, perhaps? #Resolved

File: docs/ReleaseHistory.md [hunk 2]
```diff
@@ -12,6 +12,10 @@
 - FPS => False positive reduction in static analysis.
 - FNS => False negative reduction in static analysis.
 
+# UNRELEASED
+- BRK: `SecretMasker.MaskSecrets` and `ISecretMasker.DetectSecrets` now accept an additional optional parameter. This is binary breaking only for callers, source-breaking as well for implementations of `ISecretMasker`.
+- NEW: `SecreMasker.MaskSecrets` and `ISecretMasker.DetectSecrets` now accept an callback to receive the detection results that are triggering redaction.
+
```
Comment (nguerrera) [inline docs/ReleaseHistory.md:20]: Struggled here on BRK vs. NEW since it's both so I split them.

Comment (michaelcfanning) [reply docs/ReleaseHistory.md:20]: OK, for BRK we generally want an explicit reference to name + type of added/changed thing. Also, you can use two tags for the same entry, a convention we use elsewhere.

Something like this?

- BRK, NEW: `SecretMasker.MaskSecrets` and `ISecretMasker.DetectSecrets` now accept an `Action<Detection>` callback (with a default value of null) to receive detections that result from the operation. This is binary-breaking only for `SecretMasker` callers and source-breaking as well for `ISecretMasker` implementations.

File: src/Microsoft.Security.Utilities.Core/SecretMasker.cs (hunk 1 above)
Comment (michaelcfanning) [reply src/Microsoft.Security.Utilities.Core/SecretMasker.cs]: It's not a bad name but it's true that `action` seems to be a typical term in API, so let's take your suggestion, `detectionAction`

File: src/Microsoft.Security.Utilities.Core/SecretMasker.cs [hunk 3]
```diff
@@ -142,6 +142,8 @@ public string MaskSecrets(string input)
         Detection? currentDetection = default;
         foreach (Detection detection in detections.OrderBy(x => x.Start))
         {
+            onDetection?.Invoke(detection);
```
Comment (michaelcfanning) [inline src/Microsoft.Security.Utilities.Core/SecretMasker.cs]: >onDetection [](http://example.com/codeflow?start=12&length=11)

OK, so we don't assume a need to provide any synchronization here, that would be up to action provider to handle if necessary.

You didn't create all the debt in this file, but I wonder if we should add a doc comment for this method and maybe start addressing some clean-up as we touch various things.

This file is in an unfortunately inconsistent condition but we do have some other doc xml. Maybe just add a good summary for this changed method?

Comment (michaelcfanning) [inline src/Microsoft.Security.Utilities.Core/SecretMasker.cs]: OK, so you must have thought this through and believe that firing a callback for any discrete detection (even in overlapping cases) is a good thing.

Rather than having the callbacks align precisely with masked regions. 

I tend to think that's right. It does raise the detailed point that we don't do anything explicitly related to if/when we handle the contiguous or overlapping cases. The callback receiver could do that additional work if they wanted. Just thinking aloud.

 #Resolved

File: docs/ReleaseHistory.md (hunk 2 above)
Comment (nguerrera) [reply docs/ReleaseHistory.md:20]: Ah, perfect. Thanks!

File: src/Microsoft.Security.Utilities.Core/SecretMasker.cs (hunk 3 above)
Comment (nguerrera) [reply src/Microsoft.Security.Utilities.Core/SecretMasker.cs]: Sure, I'll add a doc.

File: src/Tests.Microsoft.Security.Utilities.Core/SecretMaskerTests.cs [hunk 4]
```diff
@@ -755,6 +755,35 @@ public void SecretMasker_DistinguishLiteralAndRegexRedactionTokens()
         Assert.AreEqual("zzzxyyyab12", result);
     }
 
+    [TestMethod]
+    public void SecretMasker_ProvidesDetectionsViaCallback()
```
Comment (michaelcfanning) [inline src/Tests.Microsoft.Security.Utilities.Core/SecretMaskerTests.cs:760]: >SecretMasker_ProvidesDetectionsViaCallback [](http://example.com/codeflow?start=16&length=42)

Related to other discussion. You should consider updating this test (or adding a new one) that provokes both contiguous and overlapping matches (because in our implementation we still require the code to fire a callback for every individual match.

The regexes below would do well for this

'patter'
'pattern'
'n1'

with data like 

'yada patter pattern pattern1'

The masker should collapse this to
'yada +++ +++ +++'

But we should get six callbacks?

File: src/Microsoft.Security.Utilities.Core/SecretMasker.cs (hunk 3 above)
Comment (nguerrera) [reply src/Microsoft.Security.Utilities.Core/SecretMasker.cs]: On synchronization, yes, but now I'm again annoyed that Detection is mutable. Maybe it's time to bite that bullet. If detection were mutated here, it could break the algorithm.

Comment (nguerrera) [reply src/Microsoft.Security.Utilities.Core/SecretMasker.cs]: Yes, I did think about it, and that was my conclusion too. My immediate use case of logging what was found that triggered masking does not care about whether they were contiguous, disjoint, or overlapping.

Comment (nguerrera) [reply src/Microsoft.Security.Utilities.Core/SecretMasker.cs]: OK, I bit the bullet, see updated title description and code changes.

File: src/Tests.Microsoft.Security.Utilities.Core/SecretMaskerTests.cs (hunk 4 above)
Comment (nguerrera) [reply src/Tests.Microsoft.Security.Utilities.Core/SecretMaskerTests.cs:760]: Will do

Comment (nguerrera) [reply src/Tests.Microsoft.Security.Utilities.Core/SecretMaskerTests.cs:760]: Hmm, there's an annoying edge case here.

If detections overlap and have the same start index, then which redaction token is supposed to win? I think before my change, it was whichever detection came first, which is actually not well-defined either because it relies on hashset enumeration order. With my change, I added an unstable sort to the mix. Blah.

Maybe we should be sorting by redaction token too? I'll do that.

File: docs/ReleaseHistory.md [hunk 5]
```diff
@@ -12,6 +12,12 @@
 - FPS => False positive reduction in static analysis.
 - FNS => False negative reduction in static analysis.
 
+# UNRELEASED
+- BRK, NEW: SecretMasker.MaskSecrets and ISecretMasker.DetectSecrets now accept an Action<Detection> callback (with a default value of null) to receive detections that result from the operation. This is binary-breaking only for SecretMasker callers and source-breaking as well for ISecretMasker implementations.
```
Comment (michaelcfanning) [inline docs/ReleaseHistory.md]: >SecretMasker [](http://example.com/codeflow?start=12&length=12)

all the enquoting chars for named things are missing from this entry (though they are present for others).

SEV3 `SecretMasker.MaskSecrets`, etc.

File: docs/ReleaseHistory.md [hunk 6]
```diff
@@ -12,6 +12,12 @@
...
+- BRK: `Detection` is now sealed, and immutable. All property setters, and the copy constructor is removed.
+- BRK: `Detection` is no longer overrides Equals and GetHashCode nor implements `IEquatable<Detection>`. Equality comparison of `Detection` will now be based on reference equality.
```
Comment (michaelcfanning) [inline docs/ReleaseHistory.md]: >Equals [](http://example.com/codeflow?start=42&length=6)

`Equals` and `GetHashCode`. this standard isn't a *foolish* consistency and so it isn't a hobgoblin of a little mind. nay, it is a prudent consistency. SEV3

File: docs/ReleaseHistory.md (hunk 5 above)
Comment (nguerrera) [reply docs/ReleaseHistory.md]: I think GitHub helpfully dropped them when I cut and paste your text. Thanks, GitHub. :) Will fix.

File: src/Tests.Microsoft.Security.Utilities.Core/SecretMaskerTests.cs (hunk 4 above)
Comment (nguerrera) [reply src/Tests.Microsoft.Security.Utilities.Core/SecretMaskerTests.cs:760]: Outstanding feedback here because I broke this when I optimized how we process the list of detections. Fixed. 😊

File: src/Microsoft.Security.Utilities.Core/SecretMasker.cs [hunk 7]
```diff
@@ -122,62 +122,80 @@ public virtual void AddRegex(RegexPattern regexSecret)
...
+    /// <summary>
+    /// Masks secrets detected by <see cref="DetectSecrets(string)"/> in the
+    /// input, replacing them with <see cref="Detection.RedactionToken"/>.
+    /// </summary>
+    /// <remarks>
+    /// When secrets overlap or are adjacent to each other, the entire range
+    /// encompassing them is masked and the leftmost detection's redaction token
+    /// is used. If there is more than one overlapping leftmost detection, then
+    /// the redaction token among them that sorts first by ordinal
+    /// case-sesstring comparison is used.
```
Comment (michaelcfanning) [inline src/Microsoft.Security.Utilities.Core/SecretMasker.cs]: >case-sesstring [](http://example.com/codeflow?start=8&length=14)

? 'ordinal case-sensitive string comparison'?

**Review by michaelcfanning — APPROVED** [review]
:shipit:

File: docs/ReleaseHistory.md [hunk 8]
```diff
@@ -12,6 +12,12 @@
 - FPS => False positive reduction in static analysis.
 - FNS => False negative reduction in static analysis.
 
+# UNRELEASED
+- BRK, NEW: `SecretMasker.MaskSecrets` and `ISecretMasker.DetectSecrets` now accept an `Action<Detection>` callback (with a default value of `null`) to receive detections that result from the operation. This is binary-breaking only for `SecretMasker` callers and source-breaking as well for `ISecretMasker` implementations.
+- BRK: `Detection` is now sealed, and immutable. All property setters and the copy constructor are removed.
```
Comment (michaelcfanning) [inline docs/ReleaseHistory.md]: >, [](http://example.com/codeflow?start=32&length=1)

SEV5 you can remove this comma. Gosh! I see there is a SEV5 after all!

File: src/Microsoft.Security.Utilities.Core/SecretMasker.cs (hunk 7 above)
Comment (nguerrera) [reply src/Microsoft.Security.Utilities.Core/SecretMasker.cs]: Yikes. Fixed.

**Review by michaelcfanning — CHANGES_REQUESTED** [review]
:clock1:

**Review by michaelcfanning — APPROVED** [review]
:shipit:
